### PR TITLE
[3] feat(937): support external config

### PR DESCRIPTION
**Pull request by minzcmu**
To support one single config pipeline(with `screwdriver.yaml`) that manages a list of child pipelines, this PR adds the following stuff:
- Sync config pipeline will create/update/remove child pipelines based on `scmUrls` in screwdriver.yaml
- Sync child pipeline will getConfiguration from config pipeline; for child pipeline PRs, get Configuration from config pipeline instead of `prRef` 
- Create event will add an extra field `configPipelineSha` for child pipeline events. So that later on if ppl want to restart event for child pipeline, we know where to get the exact `screwdriver.yaml`

Related to https://github.com/screwdriver-cd/screwdriver/issues/937

### CODE REVIEW

File: lib/pipeline.js [hunk 1]
```diff
@@ -92,14 +92,31 @@ class PipelineModel extends BaseModel {
      * @param  {String}  [ref]   Reference to the branch or PR
```
Comment (Filbird) [inline lib/pipeline.js]: Can update this JSDoc :)

File: lib/pipeline.js [hunk 2]
```diff
@@ -92,14 +92,31 @@ class PipelineModel extends BaseModel {
...
      * @return {Promise}         Resolves to parsed and flattened configuration
      */
-    getConfiguration(ref) {
+    getConfiguration(configuration) {
```
Comment (Filbird) [inline lib/pipeline.js]: Is there a reason we use `configuration` over `config`?

File: lib/pipeline.js [hunk 3]
```diff
@@ -355,6 +380,158 @@ class PipelineModel extends BaseModel {
...
+        }
+        /* eslint-enable no-restricted-syntax */
+
+        return false;
+    }
+
+    /**
+     * Get a pipeline given a scmUrl
+     * @method getPipeline
+     * @param {Array} scmUrl  checkout url for a repository
```
Comment (Filbird) [inline lib/pipeline.js]: Should this be `String`?

File: lib/pipeline.js [hunk 4]
```diff
@@ -355,6 +380,158 @@ class PipelineModel extends BaseModel {
...
+            token
+        });
+
+        return scmUri;
+    }
+
+    /**
+     * Update or create a pipeline given a scmUrl
+     * @method createOrUpdatePipeine
+     * @param {Array} scmUrl  checkout url for a repository
```
Comment (Filbird) [inline lib/pipeline.js]: Same as ^

File: lib/pipeline.js [hunk 5]
```diff
@@ -355,6 +380,158 @@ class PipelineModel extends BaseModel {
...
+        const pipelineFactory = PipelineFactory.getInstance();
+        const admins = this.admins;
+        const scmContext = this.scmContext;
+        const scmUri = await this._getScmUri(scmUrl);
+        const hasAdminPermission = await this._hasAdminPermission(scmUri);
+
+        if (!hasAdminPermission) {
+            // need to figure out how to bubble up this err to user
+            // eslint-disable-next-line no-console
+            console.log(`Admins of pipeline:${this.id} has no admin permission to ${scmUrl}`);
```
Comment (Filbird) [inline lib/pipeline.js]: Perhaps this can be worded like:

"No admins of pipeline:X have admin permissions on scmUrl"

File: lib/pipeline.js [hunk 6]
```diff
@@ -355,6 +387,152 @@ class PipelineModel extends BaseModel {
...
+            // Child pipeline belongs to this parent, update it
+            if (pipeline.configPipelineId === this.id) {
+                pipeline.admins = admins;
+
+                return pipeline.update();
+            }
+
+            // Child pipline does not belong to this parent, return
+            // eslint-disable-next-line no-console
+            console.log(`Pipeline ${scmUrl} already exist: ${pipeline.id}`);
```
Comment (Filbird) [inline lib/pipeline.js]: exists

**Review by Filbird — APPROVED** [review]

File: lib/pipeline.js [hunk 7]
```diff
@@ -86,20 +87,50 @@ class PipelineModel extends BaseModel {
...
 
+    /**
+     * Get the screwdriver configuration for the pipeline at the given ref
+     * @method _getPipelineFactory
+     * @return {Object}         PipelineFactory instance
+     */
+    _getPipelineFactory() {
+        // Lazy load factory dependency to prevent circular dependency issues
+        // https://nodejs.org/api/modules.html#modules_cycles
+        /* eslint-disable global-require */
```
Comment (tkyi) [inline lib/pipeline.js]: Should either eslint-enable afterwards or use `// eslint-disable-next-line global-require`

File: lib/pipeline.js [hunk 8]
```diff
@@ -355,6 +387,152 @@ class PipelineModel extends BaseModel {
...
+            }
+        }
+        /* eslint-enable no-restricted-syntax */
+
+        return false;
+    }
+
+    /**
+     * Get a pipeline given a scmUrl
+     * @method getPipeline
```
Comment (tkyi) [inline lib/pipeline.js]: Should be `@method _getScmUri`
Also need to fix jsdoc description

File: lib/pipeline.js [hunk 9]
```diff
@@ -355,6 +387,152 @@ class PipelineModel extends BaseModel {
...
+        return scmUri;
+    }
+
+    /**
+     * Update or create a pipeline given a scmUrl
+     * @method createOrUpdatePipeine
+     * @param {String} scmUrl  checkout url for a repository
+     * @return {Promise}
+     */
+    async _createOrUpdatePipeine(scmUrl) {
```
Comment (tkyi) [inline lib/pipeline.js]: _createOrUpdatePipe`l`ine*

File: lib/pipeline.js [hunk 10]
```diff
@@ -355,6 +387,152 @@ class PipelineModel extends BaseModel {
...
+        });
+    }
+
+    /**
+     * Remove a pipeline given a scmUrl
+     * @method _removePipeine
+     * @param {String} scmUrl  checkout url for a repository
+     * @return {Promise}
+     */
+    async _removePipeine(scmUrl) {
```
Comment (tkyi) [inline lib/pipeline.js]: _removePipe`l`ine*

File: lib/eventFactory.js [hunk 11]
```diff
@@ -282,13 +306,29 @@ class EventFactory extends BaseFactory {
         return pipelineFactory.get(config.pipelineId)
             // Sync pipeline to make sure workflowGraph is generated
             .then((p) => {
+                // Sync pipeline with the parentEvent sha and create jobs based on that sha
                 if (config.parentEventId) {
                     modelConfig.parentEventId = config.parentEventId;
 
-                    // Sync pipeline with the parentEvent sha and create jobs based on that sha
+                    // for child pipelines restart event, sync with configPipelineSha
+                    if (config.configPipelineSha) {
```
Comment (tkyi) [inline lib/eventFactory.js:315]: could add jsdoc for `config.configPipelineSha`

File: lib/pipeline.js [hunk 12]
```diff
@@ -86,20 +87,50 @@ class PipelineModel extends BaseModel {
...
-        /* eslint-enable global-require */
-
         const templateFactory = TemplateFactory.getInstance();
+        const pipelineFactory = this._getPipelineFactory();
+
+        // If it is a child pipeline, use config pipeline's getConfiguration
+        if (this.configPipelineId) {
+            return pipelineFactory.get(this.configPipelineId)
+                .then((configPipeline) => {
+                    // If it is a PR on child pipeine, use config from config pipeline
```
Comment (tkyi) [inline lib/pipeline.js]: child pipe`l`ine

File: lib/eventFactory.js (hunk 11 above)
Comment (minzcmu) [reply lib/eventFactory.js:315]: Added :)

File: lib/pipeline.js [hunk 13]
```diff
@@ -86,20 +87,50 @@ class PipelineModel extends BaseModel {
         this.scmContext = config.scmContext;
     }
 
+    /**
+     * Get the screwdriver configuration for the pipeline at the given ref
```
Comment (tkyi) [inline lib/pipeline.js]: get a pipeline factory instance

File: lib/pipeline.js [hunk 14]
```diff
@@ -355,6 +387,153 @@ class PipelineModel extends BaseModel {
...
+        return scmUri;
+    }
+
+    /**
+     * Update or create a pipeline given a scmUrl
+     * @method createOrUpdatepipeline
+     * @param {String} scmUrl  checkout url for a repository
+     * @return {Promise}
+     */
+    async _createOrUpdatepipeline(scmUrl) {
```
Comment (tkyi) [inline lib/pipeline.js]: createOrUpdate`P`ipeline

File: lib/pipeline.js [hunk 15]
```diff
@@ -355,6 +387,153 @@ class PipelineModel extends BaseModel {
...
+        });
+    }
+
+    /**
+     * Remove a pipeline given a scmUrl
+     * @method _removepipeline
+     * @param {String} scmUrl  checkout url for a repository
+     * @return {Promise}
+     */
+    async _removepipeline(scmUrl) {
```
Comment (tkyi) [inline lib/pipeline.js]: remove`P`ipeline

**Review by tkyi — APPROVED** [review]
👍

**Pull request merged**